table/column.go

```diff
@@ -314,48 +314,59 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		truncateTrailingSpaces(&casted)
 	}
 
-	if v := makeStringValidator(ctx, col); v != nil {
-		str := casted.GetString()
-		strategy := charset.TruncateStrategyReplace
-		if val.Collation() == charset.CollationBin {
-			strategy = charset.TruncateStrategyTrim
-		}
-		if newStr, invalidPos := v.Truncate(str, strategy); invalidPos >= 0 {
-			casted = types.NewStringDatum(newStr)
-			err = handleWrongCharsetValue(ctx, col, str, invalidPos)
-		}
-	}
+	err = validateStringDatum(ctx, &val, &casted, col)
 	if forceIgnoreTruncate {
 		err = nil
 	}
 	return casted, err
 }
 
-func makeStringValidator(ctx sessionctx.Context, col *model.ColumnInfo) charset.StringValidator {
-	switch col.Charset {
-	case charset.CharsetASCII:
-		if ctx.GetSessionVars().SkipASCIICheck {
-			return nil
-		}
-		return charset.StringValidatorASCII{}
-	case charset.CharsetUTF8:
-		if ctx.GetSessionVars().SkipUTF8Check {
-			return nil
-		}
-		needCheckMB4 := config.GetGlobalConfig().CheckMb4ValueInUTF8
-		return charset.StringValidatorUTF8{IsUTF8MB4: false, CheckMB4ValueInUTF8: needCheckMB4}
-	case charset.CharsetUTF8MB4:
-		if ctx.GetSessionVars().SkipUTF8Check {
-			return nil
+func validateStringDatum(ctx sessionctx.Context, origin, casted *types.Datum, col *model.ColumnInfo) error {
+	// Only strings are need to validate.
+	if !(types.IsTypeVarchar(col.Tp) || types.IsTypeChar(col.Tp)) {
+		return nil
+	}
+	// Skip utf8 check if possible.
+	if mysql.IsUTF8Charset(col.Charset) && ctx.GetSessionVars().SkipUTF8Check {
+		return nil
+	}
+	// Skip ascii check if possible.
+	if col.Charset == charset.CharsetASCII && ctx.GetSessionVars().SkipASCIICheck {
+		return nil
+	}
+	// Handle string in other charsets.
+	enc := charset.FindEncoding(col.Charset)
+	if col.Charset == charset.CharsetUTF8 && config.GetGlobalConfig().CheckMb4ValueInUTF8 {
+		// Use a strict mode implementation. 4 bytes characters are invalid.
+		enc = charset.EncodingUTF8MB3StrictImpl
+	}
+	// from_binary: convert the binary to utf8.
+	if origin.Collation() == charset.CharsetBin {
+		src := casted.GetBytes()
+		encBytes, nSrc, err := enc.Decode(nil, src)
+		if err != nil {
+			var err1 error
+			// Remove the trailing invalid characters.
+			encBytes, _, err1 = enc.Decode(encBytes, src[:nSrc])
+			skipErr(err1)
```

Comment:
I think we don't need `skipErr` function.

How about the implementation below?
`encBytes, _, _ = enc.Decode(encBytes, src[:nSrc])`

Reply:
I remember the linter complains that "the error is not handled". Maybe we can improve the interface...

Reply:
I tyied it on my computer with `make lint` and `make check-static`. But no errors were found